docs/PROTOCOL.md

```diff
@@ -0,0 +1,243 @@
+    uint256 amount;      // Token amount
+}
+```
+
+### 3.4. Signature
+
+```solidity
+struct Signature {
+    uint8 v;
+    bytes32 r;
+    bytes32 s;
+}
+```
+
+### 3.5. Amount
+
+```solidity
+struct Amount {
+    address token;   // ERC-20 token address
+    uint256 amount;  // Token amount
+}
+```
+
+### 3.6. Channel Status
+
+```solidity
+enum Status {
+    VOID,     // Channel not created
+    INITIAL,  // Creation in progress
+    ACTIVE,   // Fully funded and operational
+    DISPUTE,  // Challenge period active
+    FINAL     // Ready to be closed
+}
+```
+
+### 3.7. Protocol Constants
+
+```solidity
+constant uint256 CREATOR = 0; // Creator's participant index
+constant uint256 BROKER = 1;  // Broker's participant index
+constant uint32 CHANOPEN = 7877;  // Magic number for funding state
+constant uint32 CHANCLOSE = 7879; // Magic number for closing state
+```
+
+## 4. Identifier Computation
+
+### 4.1. Channel Identifier
+
+The channelId MUST be computed as:
+
+```solidity
+channelId = keccak256(
+    abi.encode(
+        ch.participants,
+        ch.adjudicator,
+        ch.challenge,
+        ch.nonce
+    )
+);
+```
+
+### 4.2. State Hash
+
+The stateHash MUST be computed as:
+
+```solidity
+stateHash = keccak256(
+    abi.encode(
+        channelId,
+        state.data,
+        state.allocations
+    )
+);
+```
+
+For signature verification, the final hash MUST use Ethereum's personal sign message format:
+```solidity
+signatureHash = keccak256(
+    abi.encodePacked(
+        "\x19Ethereum Signed Message:\n32",
+        stateHash
+    )
+);
+```
+
+## 5. Funding Protocol
+
+### 5.1. Creation Phase
+
+1. The Creator MUST:
+   - Construct a channel configuration with participants, adjudicator, challenge period, and nonce
+   - Prepare an initial state where `state.data` MUST be set to the magic number `CHANOPEN` (7877)
+   - Define expected token deposits for all participants in the `state.allocations` array
+   - Sign the stateHash of this initial state
+   - Include their signature in the `state.sigs` array at position 0
+   - Call the `create` function with the channel configuration and initial state
+
+2. The system MUST:
+   - Verify the Creator's signature on the funding stateHash
+   - Lock the Creator's funds according to the allocation
+   - Set the channel status to `INITIAL`
+   - Emit a `Created` event with the channelId, channel configuration, and expected deposits
+
+### 5.2. Joining Phase
+
+1. Each non-Creator participant MUST:
+   - Verify the channelId and expected allocations
+   - Sign the same funding stateHash (containing the magic number `CHANOPEN`)
+   - Call the `join` function with the channelId, their participant index, and signature
+
+2. The system MUST:
+   - Verify the participant's signature against the funding stateHash
+   - Confirm the signer matches the expected participant at the given index
+   - Lock the participant's funds according to the allocation
```

Comment:
If there are different "funding" states signed singlehandedly by each participant, then ChannelImpl MUST check stateHashes are the same (meaning all allocations are the same) to AVOID situations, where some participant changed their allocation to be 0, but other participants think this party has deposited some funds.